src/main.rs

```diff
@@ -73,113 +74,140 @@ pub struct Opt {
-                    if let Some(search) = opt.search {
-                        pe.search(&search)?;
-                    } else {
-                        pe.print()?;
-                    }
-                }
-            }
-            Object::Mach(mach) => match mach {
-                mach::Mach::Fat(multi) => {
-                    for mach in &multi {
-                        match mach {
-                            Ok(binary) => {
-                                if opt.debug {
-                                    println!("{:#?}", binary);
-                                } else {
-                                    let mach = Mach(binary, opt.clone());
-                                    mach.print()?;
-                                }
-                            }
-                            Err(err) => {
-                                println!("{}", err);
-                            }
-                        }
-                    }
-                }
-                mach::Mach::Binary(binary) => {
-                    if opt.debug {
-                        println!("{:#?}", binary);
-                    } else {
-                        let mach = Mach(binary, opt.clone());
-                        mach.print()?;
-                    }
-                }
-            },
-            Object::Archive(archive) => {
-                if let Some(symbol) = opt.extract {
-                    if let Some(member) = archive.member_of_symbol(&symbol) {
-                        let bytes = archive.extract(member, &bytes)?;
-                        let mut file = File::create(Path::new(member))?;
-                        file.write_all(bytes)?;
-                    } else {
-                        return Err(anyhow::anyhow!("No member contains {:?}", symbol));
-                    }
-                } else if opt.debug {
-                    println!("archive: {:#?}", &archive);
-                } else {
-                    let archive = Archive::new(archive, opt.clone());
-                    archive.print()?;
-                }
-            }
-            _ => unreachable!(),
+
+        return Err(anyhow::anyhow!("Unknown magic: {:#x}", magic));
+    }
+
+    let object = Object::parse(&bytes)?;
+
+    // we print the semantically tagged hex table
+    if opt.hex || opt.ranges {
+        let hex = opt.hex;
+        let meta = Meta::new(&object, &bytes, opt);
+        if hex {
+            meta.print_hex()?;
+        } else {
+            meta.print_ranges()?;
         }
+        return Ok(());
+    }
+
+    // otherwise we print the kind of object
+    match object {
+        Object::Elf(elf) => parse_elf_file(opt, &bytes, elf),
+        Object::PE(pe) => parse_pe_file(opt, &bytes, pe),
+        Object::Mach(mach) => parse_mac_file(opt, mach),
+        Object::Archive(archive) => parse_archive_file(opt, &bytes, archive),
+        Object::Unknown(_magic) => unimplemented!(),
```

Comment:
let's return error (like unknown file type `magic`) in unknown case, instead of `unimplemented!()`